### PR TITLE
fix(date-utils): correctly parse date difference during cascades

### LINKED ISSUE #807: [BACKEND] [DEFECT] Set open date a week earlier for cascade creating game sessions
**Is your feature request related to a problem? Please describe.**

The game session schedule cascade creation logic is wrong, it only finds the same date of the **same** week, although this is wrong as the booking open date should be a week earlier. I.e. we get this issue:

Image

(the booking open time is later than the start time 🤡)

**BEFORE MERGING**
- [ ] Integration test written for services
- [ ] [Schemas annotated](https://payloadcms.com/docs/configuration/collections) if adding new models
- [ ] Code generation run (*hint*: `pnpm generate:types`)
- [ ] Appropriate mocks created where possible
- [ ] PR Reviewed (For non-trivial changes)
- [ ] Changes tested after rebasing on master or merging in master (*hint*: `git fetch origin master:master`, then `git rebase master` or `git merge master`)
- [ ] All required PR checks passing

### PR DESCRIPTION
**Pull request by jeffplays2005**
# Description

Fixes #807 

This PR updates and expands the test coverage and logic for date calculations related to booking open dates for sessions. The changes include:

- **Test improvements in `date.test.ts`:**
  - Corrects and clarifies test cases for `calculateOpenDate` and `getGameSessionOpenDay`.
  - Uses a consistent base date (`1970-01-01`) for `openTime` to avoid confusion and ensure time-only comparisons.
  - Adds new test cases to handle scenarios where the session and booking open day are the same, but the open time is either earlier or later than the session start time.
  - Updates expected results in tests to match the corrected logic.

- **Logic improvements in `date.ts`:**
  - Updates `getGameSessionOpenDay` to handle cases where the session and booking open day are the same:
    - If the session starts before the booking open time on the same day, the open date is set to a week earlier.
    - Otherwise, the open date is on the same day at the open time.
  - Ensures more accurate and predictable calculation of booking open dates, especially for edge cases.

These changes make the date utility functions more robust and the tests more comprehensive, reducing the risk of booking logic errors.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

- [x] Integration tests written (requires checks to pass)

## Checklist before requesting a review

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added thorough tests that prove my fix is effective and that my feature works
- [x] I've requested a review from another user

### CODE REVIEW

File: packages/shared/src/utils/date.ts [hunk 1]
```diff
@@ -66,9 +41,20 @@ export function getGameSessionOpenDay(semester: Semester, startTime: Date): Date
...
+
+  // If the session is on the same day as bookingOpenDay, check the time
+  if (daysDifference === 0) {
+    // Compare the time of startTime and bookingOpenTime (UTC)
+    const startMinutes = startTime.getUTCHours() * 60 + startTime.getUTCMinutes()
+    const openMinutes = openTime.getUTCHours() * 60 + openTime.getUTCMinutes()
+    if (startMinutes < openMinutes) {
+      // If the session starts before the open time on the same day, subtract a week
+      daysDifference = 7
+    }
```
[the remark this answers is not in the document]
Comment (jeffplays2005) [reply packages/shared/src/utils/date.ts:54]: ?
